clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -5341,6 +5341,15 @@ void CodeGenModule::EmitGlobalVarDefinition(const VarDecl *D,
       !IsDefinitionAvailableExternally &&
       D->needsDestruction(getContext()) == QualType::DK_cxx_destructor;
 
+  // It is helpless to emit the definition for an available_externally variable
+  // which can't be marked as const.
+  // We don't need to check if it needs global ctor or dtor. See the above
+  // comment for ideas.
+  if (IsDefinitionAvailableExternally &&
+      (!D->hasConstantInitialization() ||
+       !D->getType().isConstantStorage(getContext(), true, true)))
```

Comment:
Should we check for needsDestruction()?  (For example, consider a struct with a constexpr constructor, but a non-trivial destructor.)

Reply:
(we think) the available externally variables don't need ctor or dtor in the current TU.  We mentioned this in the comment.

Reply:
No, I mean, please make sure we don't emit an available_externally definition in that case.

Reply:
Got it. Makes sense. Done.